### PR TITLE
feat(ffi): Implement `RoomList` API

### PR DESCRIPTION
**Pull request by Hywan**
Address https://github.com/matrix-org/matrix-rust-sdk/issues/1911.

This patch implements the `RoomList` API inside `matrix-sdk-ffi`. It uses UniFFI and relies on the recent async UniFFI support for some methods.

This patch can be reviewed commit-by-commit.

### CODE REVIEW

File: crates/matrix-sdk-ui/src/room_list/mod.rs [hunk 1]
```diff
@@ -380,7 +373,7 @@ pub enum State {
 
     /// This state is the cruising speed, i.e. the “normal” state, where nothing
     /// fancy happens: all rooms are syncing, and life is great.
-    Enjoy,
+    CarryOn,
```
Comment (poljar) [inline crates/matrix-sdk-ui/src/room_list/mod.rs:376]: Is this a Kansas reference?

Comment (Hywan) [reply crates/matrix-sdk-ui/src/room_list/mod.rs:376]: I don't know what to say honestly :-D.

Comment (poljar) [reply crates/matrix-sdk-ui/src/room_list/mod.rs:376]: Nice :)

File: bindings/matrix-sdk-ffi/src/room_list.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    fn sync(&self) -> Arc<TaskHandle> {
+        let this = self.inner.clone();
+
+        Arc::new(TaskHandle::new(RUNTIME.spawn(async move {
+            let sync_stream = this.sync();
+            pin_mut!(sync_stream);
+
+            while let Some(_) = sync_stream.next().await {
+                // keep going!
+            }
```
[the remark this answers is not in the document]
Comment (bnjbvr) [reply bindings/matrix-sdk-ffi/src/room_list.rs:70]: Ah nevermind me, as Stefan noticed, one can listen to the state and re-create the sync loop whenever it's `Terminated`.

[the remark this answers is not in the document]
Comment (Hywan) [reply bindings/matrix-sdk-ffi/src/room_list.rs:70]: That's exactly why we have a `Terminated` state in the state-machine ;-).

**Review by jplatte — APPROVED** [review]

**Pull request merged**